nixos/profiles/bashless: init

nikstur opened the pull request:
Concludes https://github.com/NixOS/nixpkgs/issues/428908

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nikstur committed:
nixos/profiles: add bashless profile

nikstur committed:
nixos/top-level: only include preSwitchChecks when they are set

nikstur commented:
> Would there be an option to remove bash from the initrd without removing it from the full system? I think it would be reasonable to want a slimmer / more locked down early boot stage without necessarily removing behavior from the final booted system.

I think that would be a good idea. Maybe something like `boot.initrd.systemd.enableShell`.

ElvishJerricco commented on nixos/modules/config/shells-environment.nix:
```diff
@@ -36,6 +36,13 @@ in
 
   options = {
 
+    environment.enableShell = lib.mkOption {
+      default = true;
+      description = ''
+        Whether to enable a shell on your system.
+      '';
+    };
+
```
Also, use `mkEnableOption` :P As is you're missing `type`.

nikstur committed:
nixos/shadow: use /bin/sh as default userDefaultShell

The default was previously set by the bash module. It's not set directly
here (to the compiled in value) of the shadow package to allow disabling
the bash module.

Note that this doesn't mean that the effective default shell of a normal
NixOS installation has changed. It is still bash.

nikstur committed:
nixos/shells-environment: add environment.shell.enable option

nikstur committed:
nixos/systemd-initrd: add `boot.initrd.systemd.shell.enable` option

ElvishJerricco commented on nixos/tests/activation/bashless.nix:
```diff
@@ -0,0 +1,49 @@
+
+{
+
+  name = "activation-bashless";
```
I just question that this is testing anything meaningful at all. I suppose it tests that the options from the profile which *aren't* overridden by the test are bootable, but the core idea of the profile is that it's bashless and we aren't testing that *that's* bootable. I really feel like if we're going to test this, we need to test it properly.

I mean it's not like this will be a channel blocker, so it doesn't matter all that much anyway and I'm willing to be overruled on this. I just think that testing the functionality we're trying to provide is the important thing.

ElvishJerricco commented on nixos/modules/profiles/bashless.nix:
```diff
@@ -0,0 +1,46 @@
+  # Relies on bash scripts
+  powerManagement.enable = lib.mkDefault false;
+  # Has some bash inside
+  systemd.shutdownRamfs.enable = lib.mkDefault false;
```
fair enough

Ma27 reviewed (commented):
Please retarget to staging-nixos since this causes a rebuild of all nixos tests.
Everything in there will usually be merged with the next kernel update, i.e. ~weekly.

See also https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#changes-rebuilding-all-nixos-tests

nikstur committed:
nixos/test-instrumentation: add backdoor option

This allows using the test-instrumentation without the backdoor. This is
useful for example for bashless tests.

Ericson2314 commented:
Glad to see this merged! :)
